tests/pyfunc/test_pyfunc_schema_enforcement_pyspark.py

```diff
@@ -0,0 +1,323 @@
+                                [
+                                    Property("name", DataType.string),
+                                    Property("age", DataType.double),
+                                ]
+                            )
+                        ),
+                        name="array",
+                    ),
+                ]
+            ),
+        ),
+        (
+            StructType([StructField("nested_list", ArrayType(ArrayType(IntegerType())), True)]),
+            [
+                ([[1, 2, 3], [4, 5, 6], [7, 8, 9]],),
+                ([[10, 11], [12, 13, 14]],),
+            ],
+            Schema([ColSpec(Array(Array(DataType.integer)), name="nested_list")]),
+        ),
+    ],
+)
+def test_enforce_schema_spark_dataframe_complex(spark_df_schema, data, input_schema, spark):
+    input_df = spark.createDataFrame(data, spark_df_schema)
+    result = _enforce_schema(input_df, input_schema)
+    assertDataFrameEqual(input_df, result)
+
+
+def test_enforce_schema_spark_dataframe_missing_col(spark):
+    spark_df_schema = StructType(
+        [StructField("smallint", ShortType(), True), StructField("int", IntegerType(), True)]
+    )
+
+    data = [
+        (
+            1,  # smallint
+            2,  # int
+        )
+    ]
+
+    df = spark.createDataFrame(data, spark_df_schema)
+    input_schema = Schema(
+        [
+            ColSpec(DataType.integer, "smallint"),
+            ColSpec(DataType.integer, "int"),
+            ColSpec(DataType.long, "bigint"),
+        ]
+    )
+    with pytest.raises(MlflowException, match="Model is missing inputs"):
+        _enforce_schema(df, input_schema)
+
+
+def test_enforce_schema_spark_dataframe_incompatible_type(spark):
+    spark_df_schema = StructType(
+        [StructField("a", ShortType(), True), StructField("b", DoubleType(), True)]
+    )
+
+    data = [
+        (
+            1,  # a
+            2.3,  # b
+        )
+    ]
+
+    df = spark.createDataFrame(data, spark_df_schema)
+    input_schema = Schema(
+        [
+            ColSpec(DataType.integer, "a"),
+            ColSpec(DataType.integer, "b"),
+        ]
+    )
+    with pytest.raises(MlflowException, match="Incompatible input types"):
+        _enforce_schema(df, input_schema)
+
+
+def test_enforce_schema_spark_dataframe_extra_col(spark):
+    spark_df_schema = StructType(
+        [StructField("a", ShortType(), True), StructField("b", DoubleType(), True)]
+    )
+
+    data = [
+        (
+            1,  # a
+            2.3,  # b
+        )
+    ]
+
+    df = spark.createDataFrame(data, spark_df_schema)
+    input_schema = Schema([ColSpec(DataType.integer, "a")])
+    result = _enforce_schema(df, input_schema)
+    expected_result = df.drop("b")
+    assertDataFrameEqual(result, expected_result)
+
+
+def test_enforce_schema_spark_dataframe_no_schema(spark):
+    data = [
+        (
+            1,  # a
+            2.3,  # b
+        )
+    ]
+
+    df = spark.createDataFrame(data)
```

Comment:
```suggestion
    df = spark.createDataFrame(data, ["a", "b"])
```
What I would like to check is if the inferred schema mismatches for complex fields, like objects/arrays, what error would _enforce_schema throw

Reply:
Could we address this?